services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/DockerTokenUtil.java

```diff
@@ -0,0 +1,183 @@
+package org.sagebionetworks.repo.manager;
+
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.Header;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+
+import java.io.ByteArrayInputStream;
+import java.security.KeyFactory;
+import java.security.MessageDigest;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.Date;
+import java.util.List;
+
+import org.apache.commons.codec.binary.Base32;
+import org.apache.commons.net.util.Base64;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.sagebionetworks.StackConfiguration;
+
+public class DockerTokenUtil {
+
+	private static final String ISSUER = "www.synapse.org";
+	private static final long TIME_WINDOW_SEC = 120L; // two minutes
+	private static final String ACCESS = "access";
+
+	private static final String PUBLIC_KEY_ID;
+	private static final PrivateKey PRIVATE_KEY;
+	
+	// Eliptic Curve key is required by the Json Web Token signing library
+	public static final String KEY_GENERATION_ALGORITHM = "EC";
+
+	static {
+		PRIVATE_KEY = readPrivateKey();
+		X509Certificate certificate = readCertificate();
+		PUBLIC_KEY_ID = computeKeyId(certificate.getPublicKey());
+	}
+
+	public static String createToken(String userName, String type, 
+			String registry, String repository, List<String> actions, long now, String uuid) {
+
+		JSONArray access = new JSONArray();
+		JSONObject accessEntry = new JSONObject();
+
+		access.add(accessEntry);
+		accessEntry.put("type", type);
+		accessEntry.put("name", repository);
+		JSONArray actionArray = new JSONArray();
+		for (String action : actions) actionArray.add(action);
+		accessEntry.put("actions", actionArray);
+
+		Claims claims = Jwts.claims()
+				.setIssuer(ISSUER)
+				.setAudience(registry)
+				.setExpiration(new Date(now+TIME_WINDOW_SEC*1000L))
+				.setNotBefore(new Date(now-TIME_WINDOW_SEC*1000L))
+				.setIssuedAt(new Date(now))
+				.setId(uuid)
+				.setSubject(userName);
+		claims.put(ACCESS, access);
+		
+		String s = Jwts.builder().setClaims(claims).
+				setHeaderParam(Header.TYPE, Header.JWT_TYPE).
+				setHeaderParam("kid", PUBLIC_KEY_ID).
+				signWith(SignatureAlgorithm.ES256, PRIVATE_KEY).compact();
+
+		// The signature created by the Jwts library is wrong:
+		// The signature must be in P1363 format, NOT ASN.1, 
+		// which is what the code underlying 'compact()' generates when signing.
+		// Below we regenerate it.
+		try {
```

Comment:
Justify converting this way rather than using standard libraries.  Include source for this algorithm.